jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassResolver.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2018 Google LLC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.frontend;
+
+import com.google.cloud.tools.jib.builder.BuildLogger;
+import com.google.cloud.tools.jib.configuration.BuildConfiguration;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Verify;
+import javax.annotation.Nullable;
+
+/** Infers the main class in an application. */
+public class MainClassResolver {
+
+  /**
+   * If {@code mainClass} is {@code null}, tries to infer main class in this order:
+   *
+   * <ul>
+   *   <li>1. Looks in a {@code jar} plugin provided by {@code projectProperties} ({@code
+   *       maven-jar-plugin} for maven or {@code jar} task for gradle).
+   *   <li>2. Searches for a class defined with a main method.
+   * </ul>
+   *
+   * <p>Warns if main class is not valid, or throws an error if no valid main class is not found.
```

Comment:
"if no valid main class is not found" --> "if no valid main class is found"

That said, the difference between "main class is not valid" and "no valid main class is found" isn't clear to me just reading the Javadoc.

Reply:
Reworded to include "provided by projectProperties"